torch/testing/_internal/common_cuda.py

```diff
@@ -32,3 +35,41 @@ def initialize_cuda_context_rng():
         for i in range(torch.cuda.device_count()):
             torch.randn(1, device="cuda:{}".format(i))
         __cuda_ctx_rng_initialized = True
+
+
+def tf32_is_not_fp32():
+    if not torch.cuda.is_available():
+        return False
+    if torch.cuda.get_device_properties(torch.cuda.current_device()).major < 8:
+        return False
+    return True
+
+
+@contextlib.contextmanager
+def setup_tf32(dtype, rtol=0.001, atol=1e-5):
+    if dtype in {tfloat32, tcomplex64}:
+        torch.backends.cuda.matmul.allow_tf32 = True
+        if tf32_is_not_fp32():
+            yield tf32_to_fp32(dtype), rtol, atol
+        else:
+            yield tf32_to_fp32(dtype), None, None
+    else:
+        if dtype in {torch.float32, torch.complex64}:
+            torch.backends.cuda.matmul.allow_tf32 = False
+        yield dtype, None, None
+    torch.backends.cuda.matmul.allow_tf32 = True
```

Comment:
...and restore the stashed value here.

Reply:
FYI [`@contextlib.contextmanager`](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager)'s default exception handling is kind of lousy.  If code inside the context manager throws an exception, the post-yield code in the context manager actually _won't_ run.  So to clean up properly, you need the following pattern:
```python
@contextlib.contextmanager
def setup_tf32(...):
    old = torch.backends.cuda.matmul.allow_tf32
    try:
        if something:
            yield whatever
        if somethine else:
            yield whatever else
    finally:
        torch.backends.cuda.matmul.allow_tf32 = old
```

Reply:
In my opinion, this context manager is a thing that only used in test cases, and if a test case raises, then it should fail, and the whole test suit should stop, and we don't care whether a flag is restored or not.